File: ports/ts/wafproof.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { Rule, Ruleset, Counts, evaluate, Entry } from "./wafproof.js";

function ruleset(): Ruleset {
  return new Ruleset([
    { id: "xss-script", category: "xss", pattern: "<script", caseInsensitive: true },
    { id: "sqli-union", category: "sqli", pattern: "union\\s+select", caseInsensitive: true },
    { id: "pt-dotdot", category: "path-traversal", pattern: "\\.\\./" },
  ]);
}

function corpus(): Entry[] {
  return [
    { id: "m1", category: "xss", label: "malicious", text: "<script>alert(1)</script>" },
    { id: "m2", category: "sqli", label: "malicious", text: "1 UNION SELECT a,b" },
    { id: "m3", category: "path-traversal", label: "malicious", text: "../../etc/passwd" },
    { id: "b1", category: "xss", label: "benign", text: "I scripted a film" },
    { id: "b2", category: "sqli", label: "benign", text: "select a union rep" },
  ];
}

test("rule matches case-insensitively", () => {
  const rs = ruleset();
  assert.equal(rs.rules[0].matches("<SCRIPT>x"), true);
  assert.equal(rs.rules[2].matches("assets/logo.png"), false);
});

test("invalid regex throws", () => {
  assert.throws(() => new Rule({ id: "bad", pattern: "(" }));
});

test("detect fires on any rule", () => {
  const rs = ruleset();
  assert.equal(rs.detect("hello ../../x"), true);
  assert.equal(rs.detect("perfectly benign text"), false);
});

test("evaluate counts", () => {
  const rs = ruleset();
  const c = evaluate((t) => rs.detect(t), corpus());
  assert.equal(c.tp, 3);
  assert.equal(c.fn, 0);
  assert.equal(c.tn, 2);
  assert.equal(c.total, 5);
});

test("metrics perfect", () => {
  const c = new Counts();
  c.tp = 3;
  c.tn = 2;
  assert.ok(Math.abs(c.recall - 1.0) < 1e-9);
  assert.ok(Math.abs(c.precision - 1.0) < 1e-9);
  assert.ok(Math.abs(c.f1 - 1.0) < 1e-9);
  assert.ok(Math.abs(c.falsePositiveRate - 0.0) < 1e-9);
});

test("metrics safe-div on empty", () => {
  const c = new Counts();
  assert.equal(c.recall, 0.0);
  assert.equal(c.precision, 0.0);
  assert.equal(c.f1, 0.0);
});

test("overbroad rule lowers precision", () => {
  const rs = new Ruleset([{ id: "broad", category: "sqli", pattern: "select", caseInsensitive: true }]);
  const c = evaluate((t) => rs.detect(t), corpus());
  assert.ok(c.fp >= 1);
  assert.ok(c.precision < 1.0);
});

test("matching attribution", () => {
  const rs = ruleset();
  const ids = rs.matching("<script>../../x").map((r) => r.id);
  assert.ok(ids.includes("xss-script"));
  assert.ok(ids.includes("pt-dotdot"));
});

File: ports/go/wafproof_test.go
package wafproof

import (
	"math"
	"testing"
)

func sampleRuleset(t *testing.T) Ruleset {
	rs := Ruleset{
		{ID: "xss-script", Category: "xss", Pattern: "<script", CaseInsensitive: true},
		{ID: "sqli-union", Category: "sqli", Pattern: `union\s+select`, CaseInsensitive: true},
		{ID: "pt-dotdot", Category: "path-traversal", Pattern: `\.\./`},
	}
	if err := rs.Compile(); err != nil {
		t.Fatalf("compile failed: %v", err)
	}
	return rs
}

func sampleCorpus() []Entry {
	return []Entry{
		{ID: "m1", Category: "xss", Label: "malicious", Text: "<script>alert(1)</script>"},
		{ID: "m2", Category: "sqli", Label: "malicious", Text: "1 UNION SELECT a,b"},
		{ID: "m3", Category: "path-traversal", Label: "malicious", Text: "../../etc/passwd"},
		{ID: "b1", Category: "xss", Label: "benign", Text: "I scripted a film"},
		{ID: "b2", Category: "sqli", Label: "benign", Text: "select a union rep"},
	}
}

func TestRuleMatches(t *testing.T) {
	rs := sampleRuleset(t)
	if !rs[0].Matches("<SCRIPT>x") {
		t.Error("case-insensitive script rule should match <SCRIPT>")
	}
	if rs[2].Matches("assets/logo.png") {
		t.Error("path rule should not match a clean path")
	}
}

func TestCompileInvalidRegex(t *testing.T) {
	rs := Ruleset{{ID: "bad", Pattern: "("}}
	if err := rs.Compile(); err == nil {
		t.Error("expected compile error for invalid regex")
	}
}

func TestDetectAny(t *testing.T) {
	rs := sampleRuleset(t)
	if !rs.Detect("hello ../../x") {
		t.Error("detect should fire on traversal")
	}
	if rs.Detect("perfectly benign text") {
		t.Error("detect should not fire on benign text")
	}
}

func TestEvaluateCounts(t *testing.T) {
	rs := sampleRuleset(t)
	c := Evaluate(rs.Detect, sampleCorpus())
	if c.TP != 3 {
		t.Errorf("TP = %d, want 3", c.TP)
	}
	if c.FN != 0 {
		t.Errorf("FN = %d, want 0", c.FN)
	}
	if c.TN != 2 {
		t.Errorf("TN = %d, want 2", c.TN)
	}
	if c.Total() != 5 {
		t.Errorf("total = %d, want 5", c.Total())
	}
}

func almost(a, b float64) bool { return math.Abs(a-b) < 1e-9 }

func TestMetrics(t *testing.T) {
	c := Counts{TP: 3, FP: 0, FN: 0, TN: 2}
	if !almost(c.Recall(), 1.0) {
		t.Errorf("recall = %v, want 1.0", c.Recall())
	}
	if !almost(c.Precision(), 1.0) {
		t.Errorf("precision = %v, want 1.0", c.Precision())
	}
	if !almost(c.F1(), 1.0) {
		t.Errorf("f1 = %v, want 1.0", c.F1())
	}
	if !almost(c.FalsePositiveRate(), 0.0) {
		t.Errorf("fpr = %v, want 0.0", c.FalsePositiveRate())
	}
}

func TestMetricsSafeDiv(t *testing.T) {
	var c Counts
	if c.Recall() != 0.0 || c.Precision() != 0.0 || c.F1() != 0.0 {
		t.Error("empty counts should yield 0.0 metrics, not NaN")
	}
}

func TestOverbroadLowersPrecision(t *testing.T) {
	rs := Ruleset{{ID: "broad", Category: "sqli", Pattern: "(?i)select"}}
	if err := rs.Compile(); err != nil {
		t.Fatal(err)
	}
	c := Evaluate(rs.Detect, sampleCorpus())
	// "select" matches malicious m2 and benign b2 -> a false positive
	if c.FP == 0 {
		t.Error("overbroad rule should produce at least one FP")
	}
	if c.Precision() >= 1.0 {
		t.Error("precision should drop below 1.0 with an FP")
	}
}
